ext/ddtrace_profiling_native_extension/http_transport.c

```diff
@@ -0,0 +1,281 @@
+#include <ruby.h>
+#include <ruby/thread.h>
+#include <ddprof/ffi.h>
+
+// Used to report profiling data to Datadog.
+// This file implements the native bits of the Datadog::Profiling::HttpTransport class
+
+static VALUE ok_symbol = Qnil; // :ok in Ruby
+static VALUE error_symbol = Qnil; // :error in Ruby
+
+static ID agentless_id; // id of :agentless in Ruby
+static ID agent_id; // id of :agent in Ruby
+
+#define byte_slice_from_literal(string) ((ddprof_ffi_ByteSlice) {.ptr = (uint8_t *) "" string, .len = sizeof("" string) - 1})
+
+struct call_exporter_without_gvl_arguments {
+  ddprof_ffi_ProfileExporterV3 *exporter;
+  ddprof_ffi_Request *request;
+  ddprof_ffi_SendResult result;
+};
+
+inline static ddprof_ffi_ByteSlice byte_slice_from_ruby_string(VALUE string);
+static VALUE _native_validate_exporter(VALUE self, VALUE exporter_configuration);
+static ddprof_ffi_NewProfileExporterV3Result create_exporter(VALUE exporter_configuration, VALUE tags_as_array);
+static VALUE handle_exporter_failure(ddprof_ffi_NewProfileExporterV3Result exporter_result);
+static ddprof_ffi_EndpointV3 endpoint_from(VALUE exporter_configuration);
+static void convert_tags(ddprof_ffi_Tag *converted_tags, long tags_count, VALUE tags_as_array);
+static VALUE _native_do_export(
+  VALUE self,
+  VALUE exporter_configuration,
+  VALUE upload_timeout_milliseconds,
+  VALUE start_timespec_seconds,
+  VALUE start_timespec_nanoseconds,
+  VALUE finish_timespec_seconds,
+  VALUE finish_timespec_nanoseconds,
+  VALUE pprof_file_name,
+  VALUE pprof_data,
+  VALUE code_provenance_file_name,
+  VALUE code_provenance_data,
+  VALUE tags_as_array
+);
+static ddprof_ffi_Request *build_request(
+  ddprof_ffi_ProfileExporterV3 *exporter,
+  VALUE upload_timeout_milliseconds,
+  VALUE start_timespec_seconds,
+  VALUE start_timespec_nanoseconds,
+  VALUE finish_timespec_seconds,
+  VALUE finish_timespec_nanoseconds,
+  VALUE pprof_file_name,
+  VALUE pprof_data,
+  VALUE code_provenance_file_name,
+  VALUE code_provenance_data
+);
+static void *call_exporter_without_gvl(void *exporter_and_request);
+
+void http_transport_init(VALUE profiling_module) {
+  VALUE http_transport_class = rb_define_class_under(profiling_module, "HttpTransport", rb_cObject);
+
+  rb_define_singleton_method(http_transport_class, "_native_validate_exporter",  _native_validate_exporter, 1);
+  rb_define_singleton_method(http_transport_class, "_native_do_export",  _native_do_export, 11);
+
+  ok_symbol = ID2SYM(rb_intern_const("ok"));
+  error_symbol = ID2SYM(rb_intern_const("error"));
+  agentless_id = rb_intern_const("agentless");
+  agent_id = rb_intern_const("agent");
+}
+
+inline static ddprof_ffi_ByteSlice byte_slice_from_ruby_string(VALUE string) {
+  Check_Type(string, T_STRING);
+  ddprof_ffi_ByteSlice byte_slice = {.ptr = (uint8_t *) StringValuePtr(string), .len = RSTRING_LEN(string)};
```

Comment:
How is the memory pointed by the pointer returned by `StringValuePtr` managed? Does it point to a GCed memory area, or is it a raw pointer? If a raw pointer, how do we make sure it doesn't leak (especially regarding the use of `rb_raise`)?

Reply:
It's a raw pointer to the actual `char *` that Ruby strings keep. So it has the same lifetime as the Ruby object, and we don't need to clean it up, since the Ruby GC will manage its lifetime (when the object gets collected, the string will get free'd).

On the other hand, we do have to be careful around concurrency, e.g. we must never manipulate Ruby strings if we have released the Global VM Lock, because Ruby may decide to move/change/collect them concurrently.